#figs for manuscript
---
title: "Dolen_OT_neurons_figs"
author: "Genevieve Stein-O'Brien"
date: "3/15/2019"
output: html_document
---

# Initialize Environment
```{r init}
source('code/init.R')
library(reticulate)
use_condaenv('OT-reticulate')
set.seed(1234)
library("AnnotationDbi")
library("org.Mm.eg.db")
library("biomaRt")
library("RColorBrewer")
library(corrplot)
library(monocle)
library("yarrr")
```



```{r load_data}
dat<-readRDS(file="../data/filtered10kOxy.rds")

# Fix levels on source_plate
pData(dat)$source_plate<-factor(pData(dat)$source_plate,levels=c("Plate 1 06/19/18"," Plate 2 05/17/18","Plate 3 05/17/18"))

#call cell types by clusters
pData(dat)$CellType<-"Magnocellular"
pData(dat)$CellType[pData(dat)$Cluster==1]<-"Parvocellular"
PMcolors<-c("orangered","darkcyan")

PV_vs_Magno_res<-readRDS("PV_vs_Magno_res.rds")
nCellCutoff<-5
expressed_genes<-rownames(fData(dat))[fData(dat)$num_cells_expressed>=nCellCutoff]

#qval_cutoff<-0.0000001
qval_cutoff<-0.001
PV_vs_Magno_sigGeneIDs<-PV_vs_Magno_res$gene_id[PV_vs_Magno_res$qval<=qval_cutoff]
length(PV_vs_Magno_sigGeneIDs)
```

```{r seq_summary_stats}
seq.dat<-read.csv("../data/HGCWGBCX2.demux-report.csv")

seq.dat %>%
  filter(Expected == "Expected") %>%
  group_by(Index) %>%
  dplyr::summarize(totalCount = sum(Count)) %>%
  dplyr::summarize(mean = mean(totalCount,na.rm=TRUE),sd=sd(totalCount,na.rm=TRUE),n=n(),se=sd/sqrt(n))
  
```


```{r fig5bc}
celltype.p<-plot_cell_clusters_loyal(dat,color="CellType",cell_size=2) + scale_color_manual(values=PMcolors) + theme(legend.position = c(0.2, 0.8)) + coord_equal(1) + labs(tag = "B")
fluorogold.p<-plot_cell_clusters_loyal(dat,color="Fluorogold",cell_size=2) + scale_color_manual(values=label_colors) + theme(legend.position = c(0.2, 0.8)) + coord_equal(1) + labs(tag="C")

pdf("../plots/fig5bc.pdf",width=10,height=5,fonts="Helvetica")
grid.arrange(
celltype.p,
fluorogold.p,
	ncol=2,nrow=1
)
dev.off()


   p<-ggplot(tmp,aes(x=tSNE1_pos,y=tSNE2_pos)) + 
   	  geom_point(aes_string(color=color_by,alpha="value"),stroke=0,size=cell_size) + 
      facet_wrap('gene_short_name')+ theme_bw() + scale_color_manual(values=PMcolors) + 
      monocle:::monocle_theme_opts() + scale_alpha(range=c(0.05,1)) 



pData(dat)$Fluorogold
pData(dat)$Cluster
sort_date
source_plate


 
#coordinates for plotting 
dat@reducedDimS


```

```{r fig5d}

replicates.p <- ggplot(pData(dat)) + 
  geom_bar(aes(x=interaction(source_plate,CellType),fill=Fluorogold),position="fill") +
  scale_fill_manual(values=label_colors) + 
  ylab("Proportion of total cells") + 
  theme(axis.text.x=element_text(angle=-90,hjust=1)) +
  monocle:::monocle_theme_opts()

replicates.p

pdf("../plots/fig5d-FG_label_by_celltype_and_replicate.pdf",width=5,height=6)
replicates.p
dev.off()

#distribution of cell types by replicate



```

```{r fig5e}

replicates_nGenes.p <- ggplot(pData(dat)) + 
  geom_boxplot(aes(x=interaction(CellType,source_plate),y=num_genes_expressed,fill=CellType)) +
  scale_fill_manual(values=PMcolors) + 
  ylab("Number of Genes Detected") + 
  theme(axis.text.x=element_text(angle=-90,hjust=1)) +
  monocle:::monocle_theme_opts() + 
  theme(legend.position = c(0.5, 0.2))

replicates_nGenes.p

pdf("../plots/fig5e-nGenes_expressed_by_celltype_and_replicate.pdf",width=5,height=6)
replicates_nGenes.p
dev.off()

```



```{r fig5f_variance_test}

parvo.dis<-dist(t(exprs(dat[expressed_genes,pData(dat)$CellType=='Parvocellular'])))
parvo.distData<-data.frame(celltype="Parvocellular",JSD=as.vector(parvo.dis))

magno.dis<-dist(t(exprs(dat[expressed_genes,pData(dat)$CellType=='Magnocellular'])))
magno.distData<-data.frame(celltype="Magnocellular",JSD=as.vector(magno.dis))

#rand.dis<-dist(t(exprs(dat[expressed_genes,rownames(pData(dat)) %in% sample(rownames(pData(dat)),30)])))

rand.dis<-vector()
for(i in 1:10){
temp<-dist(t(exprs(dat[expressed_genes,rownames(pData(dat)) %in% sample(rownames(pData(dat)),21)])))
rand.dis<-c(rand.dis,temp)
}
#rand.dis<-dist(t(exprs(dat[expressed_genes,rownames(pData(dat))])))
rand.distData<-data.frame(celltype="rand",JSD=as.vector(rand.dis))
rand.distData.ecdf<-ecdf(rand.distData$JSD)

distData<-rbind(parvo.distData,magno.distData,rand.distData)


# DKW confidence intervals
n<-200
alpha<-0.05
eps<-sqrt(log(2/alpha)/(2*n))
#xx            <-    seq(60,120,length.out=1000)
xx      <-  knots(rand.distData.ecdf)
ll            <-    pmax(rand.distData.ecdf(xx)-eps,0)
uu             <-    pmin(rand.distData.ecdf(xx)+eps,1)
randCI<-as.data.frame(cbind(xx,ll,uu))

# ks test
ks.test(parvo.distData$JSD,magno.distData$JSD,exact=T)

# Difference in means
t.test(parvo.distData$JSD,magno.distData$JSD)

#Variance test
var.test(parvo.distData$JSD,magno.distData$JSD,alternative="greater")

distData<-rbind(parvo.distData,magno.distData)

var_test.p<-ggplot(distData) + 
  #geom_ribbon(aes(x=xx,ymin=ll,ymax=uu),color="grey80",fill="grey80",alpha=0.25,data=randCI) + 
  stat_ecdf(aes(x=JSD,color=celltype),geom="line") + 
  monocle:::monocle_theme_opts() + 
  scale_color_manual(values=c(PMcolors)) + 
  ggtitle("Intra-celltype heterogeneity of gene expression") + 
  ylab("ecdf") + xlab("Euclidean Distance") +
  geom_hline(yintercept=1,linetype="dashed",color="grey70") + 
  geom_hline(yintercept=0,linetype="dashed",color="grey70") +
  theme(legend.position = c(0.8, 0.2))
var_test.p

pdf("../plots/ecdf.pdf",width=4,height=4)
var_test.p + theme(legend.position = c(0.8, 0.5))
dev.off()

```



```{r fig6a_heatmap}

myPHeatmap<-function(cds,geneset,logMode=TRUE){
  sub<-cds[lookupGeneId(cds,geneset),]
  mat<-as.matrix(exprs(sub))
  if(logMode){
    mat<-log10(mat+1)
  }
  pheatmap(mat=log10(as.matrix(exprs(sub))+1),
            scale="row",
            labels_row=fData(sub)$gene_short_name,
            annotation_col=pData(sub)[,c("source_plate","Total_mRNAs","num_genes_expressed","Cluster","Fluorogold")],
            labelsCol=FALSE,  
            #annotation_colors=c(label_colors,PMcolors),
            #color = colorRampPalette(piratepal(palette="brave"))(100), #rev(brewer.pal(n = 11, name ="RdBu")))(100),
            #color = magma(100),
            #color = colorRampPalette(c("steelblue", "white", "firebrick3"))(100),
            color = colorRampPalette(c("aquamarine4", "white", "darkorange4"))(100),
            clustering_distance_cols = "canberra",
            clustering_distance_rows = "correlation",
            show_rownames = T, show_colnames = F,
            cutree_rows=2, cutree_cols=2, breaks=seq(-3,3,length=101))
}
myPHeatmap(dat,PV_vs_Magno_sigGeneIDs,logMode=TRUE)

HM<-myPHeatmap(dat,PV_vs_Magno_sigGeneIDs,logMode=TRUE)

pdf("../plots/Parvo_vs_Magno_sigGenes_heatmap.pdf",width=8,height=10)
HM
dev.off()
```

```{r fig6bc}

genes2plot <- dat[row.names(subset(fData(dat), gene_short_name %in% c("Kcnmb4","Calb1","Cnr1","Reln"))),]

genes_order<-c("Calb1","Kcnmb4","Reln","Cnr1")
#genes_order<-c("Bdnf","Esr2","Crh","Penk")


key_genes_umap.p<-plot_cell_clusters(dat,markers=genes_order,ncol=1,cell_size=2) + 
		scale_color_viridis(option="viridis") + scale_size_area(max_size = 6) + coord_equal(1) + guides(color=FALSE)
key_genes_violin.p<-monocle::plot_genes_violin(genes2plot, grouping="CellType", ncol=1, min_expr=0.1,
		color_by="CellType",panel_order=genes_order) +
		scale_fill_manual(values=PMcolors) + theme(legend.position = "top") + monocle:::monocle_theme_opts() + guides(fill=FALSE)


pdf("../plots/fig6bc.pdf",width=8,height=10,fonts="Helvetica")
grid.arrange(
	key_genes_umap.p,
	key_genes_violin.p,
	ncol=2,nrow=4, widths = c(1, 0.75), #heights=c(4,4,4,4),
  layout_matrix = rbind(c(1,2),
                        c(1,2),
                        c(1,2),
                        c(1,2))
)
dev.off()


key_genes_umap.p<-plot_cell_clusters(dat,markers=genes_order,ncol=1,cell_size=2) + 
		scale_color_viridis(option="viridis") + scale_size_area(max_size = 6) + coord_equal(1)

legend <- cowplot::get_legend(key_genes_umap.p) 

pdf("../plots/umap_legend_fig6.pdf",width=5,height=5)
grid.draw(legend) 
dev.off()
```


```{r density_plots}
# Get gene sets
ASDRiskGenes<-readxl::read_excel("../annotations/ASDRiskGenes.xlsx",sheet="FMRP ASD Risk Genes",skip=1)
ASDRiskGenes
ASDRiskGenes<-ASDRiskGenes[ASDRiskGenes$Gene %in% lookupGeneName(dat,expressed_genes),]

NewASDRiskGenes<-readxl::read_excel("../annotations/ASDRiskGenes_Ale_update_6-5-19.xlsx",sheet="ASD Risk Genes",skip=1)
NewASDRiskGenes$Gene<-tools::toTitleCase(stringr::str_to_lower(NewASDRiskGenes$Gene))
NewASDRiskGenes
NewASDRiskGenes<-NewASDRiskGenes[NewASDRiskGenes$Gene %in% lookupGeneName(dat,expressed_genes),]
categoryCutoff<-4
NewASDRiskGenes<-NewASDRiskGenes[NewASDRiskGenes$Category<=categoryCutoff,]

SFARIRiskGenes<-readxl::read_excel("../annotations/ASDRiskGenes.xlsx",sheet="SFARI ASD RISK GENES",skip=2)
SFARIRiskGenes
SFARIRiskGenes$mouseGeneName<-tools::toTitleCase(stringr::str_to_lower(SFARIRiskGenes$Gene))
SFARIRiskGenes<-SFARIRiskGenes[SFARIRiskGenes$mouseGeneName %in% lookupGeneName(dat,expressed_genes),]

FMRPgenes<-readxl::read_excel("../annotations/FMRP binding partners Darnell.xlsx")
FMRPgenes
FMRPgenes<-FMRPgenes[FMRPgenes$'gene symbol' %in% lookupGeneName(dat,expressed_genes),]

#use homologene for SFARIR
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")

map<-getLDS(attributes = c("ensembl_gene_id","hgnc_symbol"),filters = "hgnc_symbol", values = SFARIRiskGenes$Gene, mart = human,
      attributesL = c("ensembl_gene_id","mgi_symbol","chromosome_name","start_position"), martL = mouse)
str(map)

indx<-match(SFARIRiskGenes$Gene,map$HGNC.symbol )
SFARIRiskGenes$Gene[1]; map$HGNC.symbol[indx[1]]
indx<-indx[!is.na(indx)]
map<-map[indx,]             
map<-map[map$MGI.symbol %in% lookupGeneName(dat,expressed_genes),]


#Calculate Cell Type Specific mean expression
fData(dat)$Magno_mean_cpc<-Matrix::rowMeans(exprs(dat)[,pData(dat)$CellType=="Magnocellular"])
fData(dat)$Parvo_mean_cpc<-Matrix::rowMeans(exprs(dat)[,pData(dat)$CellType=="Parvocellular"])
fData(dat)$logfc<-log2(fData(dat)$Magno_mean_cpc/fData(dat)$Parvo_mean_cpc)

fData(dat[lookupGeneId(dat,intersect(ASDRiskGenes$Gene,lookupGeneName(dat,PV_vs_Magno_sigGeneIDs))),])$logfc
fData(dat[lookupGeneId(dat,intersect(map$MGI.symbol,lookupGeneName(dat,PV_vs_Magno_sigGeneIDs))),])$logfc
fData(dat[lookupGeneId(dat,intersect(FMRPgenes$'gene symbol',lookupGeneName(dat,PV_vs_Magno_sigGeneIDs))),])$logfc


#fData(dat)$ASDRiskGene<-fData(dat)$gene_short_name %in% intersect(FMRPgenes$'gene symbol', map$MGI.symbol)
fData(dat)$NewASDRiskGene<-fData(dat)$gene_short_name %in% NewASDRiskGenes$Gene # What is this for?
fData(dat)$SFARIRiskGene<-fData(dat)$gene_short_name %in% SFARIRiskGenes$mouseGeneName
fData(dat)$SFARIRiskGene<-fData(dat)$gene_short_name %in% map$MGI.symbol
fData(dat)$FMRPgenes<-fData(dat)$gene_short_name %in% FMRPgenes$'gene symbol'
fData(dat)$isSignificant<-fData(dat)$gene_short_name %in% lookupGeneName(dat,PV_vs_Magno_sigGeneIDs)



p<-ggplot(fData(dat[expressed_genes,])) +
  geom_density(aes(x=logfc),alpha=0,color="black") + 
  #geom_jitter(aes(x=logfc,y=0.3),height=0.02,color="blue",data=fData(dat)[fData(dat)$ASDRiskGene,]) + 
  geom_density(aes(x=logfc),color="orange",fill="orange",alpha=0.3,data=fData(dat)[fData(dat)$NewASDRiskGene,]) + 
  #geom_rug(aes(x=logfc),color="orange",fill="orange",alpha=0.3,data=fData(dat)[fData(dat)$ASDRiskGene,]) + 

  #geom_text_repel(aes(x=logfc,y=0,label=gene_short_name),color="black",data=fData(dat)[fData(dat)$ASDRiskGene,]) + 
  #geom_jitter(aes(x=logfc,y=0.1),height=0.02,color="red",data=fData(dat)[fData(dat)$ASDRiskGene & fData(dat)$isSignificant,]) + 
  geom_density(aes(x=logfc),color="red",fill="red",alpha=0.3,data=fData(dat)[fData(dat)$NewASDRiskGene & fData(dat)$isSignificant,]) + 
  #geom_density(aes(x=logfc),color="cyan",fill="cyan",alpha=0.3,data=fData(dat)[fData(dat)$FMRPgenes,]) + 
  #geom_density(aes(x=logfc),color="blue",fill="blue",alpha=0.3,data=fData(dat)[fData(dat)$FMRPgenes & fData(dat)$isSignificant,]) + 
  #geom_jitter(aes(x=logfc,y=0.3),height=0.02,color="green",data=fData(dat)[fData(dat)$SFARIRiskGene,]) + 
  #geom_density(aes(x=logfc),color="green",fill="green",alpha=0.3,data=fData(dat)[fData(dat)$SFARIRiskGene,]) + 
  #geom_jitter(aes(x=logfc,y=0.1),height=0.02,color="red",data=fData(dat)[fData(dat)$SFARIRiskGene & fData(dat)$isSignificant,]) + 
  #geom_density(aes(x=logfc),color="purple",fill="purple",alpha=0.3,data=fData(dat)[fData(dat)$SFARIRiskGene & fData(dat)$isSignificant,]) + 
  
  #geom_point(aes(x=logfc,y=0),color="red",data=fData(dat)[fData(dat)$ASDRiskGene & fData(dat)$isSignificant,]) + 
  geom_label_repel(aes(x=logfc,y=0,label=gene_short_name),color="black",data=fData(dat)[fData(dat)$NewASDRiskGene & fData(dat)$isSignificant,]) + 

  geom_rug(aes(x=logfc),color="orange",fill="orange",alpha=0.3,data=fData(dat)[fData(dat)$NewASDRiskGene,]) + 
  #geom_rug(aes(x=logfc),color="cyan",fill="cyan",alpha=0.3,data=fData(dat)[fData(dat)$FMRPgenes,]) +

  xlim(-10,10)+
  geom_vline(aes(xintercept=0),linetype="dashed") +
  monocle:::monocle_theme_opts()
p

pdf("../plots/logfc_distribution_by_genessets.pdf",width=5,height=5)
p
dev.off()

Data(dat)[fData(dat)$FMRPgenes & fData(dat)$isSignificant,],binwidth=.5)

p<-ggplot(fData(dat[expressed_genes,])) +
  geom_density(aes(x=logfc),alpha=0,color="black") + 
  geom_density(aes(x=logfc),color="orange",fill="orange",alpha=0.3,data=fData(dat)[fData(dat)$NewASDRiskGene,]) + 
  geom_density(aes(x=logfc),color="red",fill="red",alpha=0.3,data=fData(dat)[fData(dat)$NewASDRiskGene & fData(dat)$isSignificant,]) +
  geom_vline(aes(xintercept=0),linetype="dashed") +
  xlim(-10,10)+ ggtitle("Density of ASD Risk Genes (<=Cat4)") +
  monocle:::monocle_theme_opts()
p

pdf("../plots/logfc_distribution_by_FMRP-ASDRiskGene.pdf",width=5,height=5)
p
dev.off()

p<-ggplot(fData(dat[expressed_genes,])) +
  #geom_histogram(aes(x=logfc),alpha=0,color="black") +
  geom_histogram(aes(x=logfc,y=..density..),color="cyan",fill="cyan",alpha=0.3,data=fData(dat)[fData(dat)$FMRPgenes,]) + 
  geom_histogram(aes(x=logfc,y=..density..),color="blue",fill="blue",alpha=0.3,data=fData(dat)[fData(dat)$FMRPgenes & fData(dat)$isSignificant,],binwidth=.5) + 
   xlim(-10,10)+  ggtitle("Density of FMRP Genes") +
  geom_vline(aes(xintercept=0),linetype="dashed") +
  monocle:::monocle_theme_opts()
p

pdf("../plots/logfc_distribution_by_FMRP.pdf",width=5,height=5)
p
dev.off()

p<-ggplot(fData(dat[expressed_genes,])) +
  geom_density(aes(x=logfc),alpha=0,color="black") +
  geom_density(aes(x=logfc),color="cyan",fill="cyan",alpha=0.3,data=fData(dat)[fData(dat)$FMRPgenes,]) + 
  geom_density(aes(x=logfc),color="blue",fill="blue",alpha=0.3,data=fData(dat)[fData(dat)$FMRPgenes & fData(dat)$isSignificant,]) + 
  #geom_jitter(aes(x=logfc,y=0.3),height=0.02,color="green",data=fData(dat)[fData(dat)$SFARIRiskGene,]) + 
  geom_density(aes(x=logfc),color="green",fill="green",alpha=0.3,data=fData(dat)[fData(dat)$SFARIRiskGene,]) + 
  #geom_jitter(aes(x=logfc,y=0.1),height=0.02,color="red",data=fData(dat)[fData(dat)$SFARIRiskGene & fData(dat)$isSignificant,]) + 
  geom_density(aes(x=logfc),color="purple",fill="purple",alpha=0.3,data=fData(dat)[fData(dat)$SFARIRiskGene & fData(dat)$isSignificant,]) + 
  xlim(-10,10)+
  geom_vline(aes(xintercept=0),linetype="dashed") +
  monocle:::monocle_theme_opts()
p

pdf("../plots/logfc_distribution_by_genessets.pdf",width=5,height=5)
p
dev.off()




```



```{r ref}
myTSNEPlotAlpha<-function(cds,markers=NULL,logMode=T,color_by="color",shape_by=NULL,scaled=FALSE,cell_size=2){
  tmp<-pData(cds)
  if(!is.null(markers)){
    genes<-as.matrix(exprs(cds[rownames(fData(cds)) %in% lookupGeneId(cds,markers)]))
    if(logMode){
      genes<-log10(genes+1)
    }
    geneMeans<-rowMax(genes)
    if(scaled){
      genes<-genes/geneMeans
    }
    genes<-t(genes)
    genes<-melt(genes)
    colnames(genes)<-c("cell_id","gene_id","value")
    genes<-merge(genes,fData(cds),by.x="gene_id",by.y="gene_id",all.x=TRUE,sort=FALSE)
    tmp<-merge(tmp,genes,by.x=0,by.y="cell_id")
    p<-ggplot(tmp,aes(x=tSNE1_pos,y=tSNE2_pos))
    if(is.null(shape_by)){
      p + geom_point(aes_string(color=color_by,alpha="value"),stroke=0,size=cell_size) + facet_wrap('gene_short_name')+ theme_bw() + scale_color_brewer(palette="Set1") + monocle:::monocle_theme_opts() + scale_alpha(range=c(0.05,1)) 
    }else{
      p + geom_point(aes_string(color=color_by,alpha="value",stroke=0,shape=shape_by),size=cell_size) + facet_wrap('gene_short_name')+ theme_bw() + scale_color_brewer(palette="Set1")+ monocle:::monocle_theme_opts() + scale_alpha(range=c(0.05,1)) 
    }
  }else{
    p<-ggplot(tmp,aes(x=tSNE1_pos,y=tSNE2_pos))
    if(is.null(shape_by)){
      p + geom_point(aes_string(color=color_by),size=cell_size) + theme_bw() + scale_color_brewer(palette="Set1")+ monocle:::monocle_theme_opts() 
    }else{
      p + geom_point(aes_string(color=color_by,shape=shape_by),size=cell_size) + theme_bw() + scale_color_brewer(palette="Set1")+ monocle:::monocle_theme_opts() 
    }
  }
}
```

```{r}
plot_cell_clusters_loyal<-function (cds, x = 1, y = 2, color_by = "Cluster", markers = NULL, 
    show_cell_names = FALSE, cell_size = 1.5, cell_name_size = 2, 
    min_expr = 0.1, show_group_id = FALSE, nrow = NULL, ncol = NULL, 
    ...) 
{
    plotting_func = ggplot2::geom_point
    if (length(pData(cds)$Cluster) == 0) {
        stop("Error: Clustering is not performed yet. Please call clusterCells() before calling this function.")
    }
    if (cds@dim_reduce_type == "tSNE") {
        low_dim_coords <- cds@reducedDimA
    }
    else {
        low_dim_coords <- cds@reducedDimS
    }
    if (nrow(low_dim_coords) == 0) {
        message("reduceDimension is not performed yet. We are plotting the normalized reduced space obtained from preprocessCDS function.")
        low_dim_coords <- t(cds@normalized_data_projection)
    }
    gene_short_name <- NULL
    sample_name <- NULL
    data_dim_1 <- NULL
    data_dim_2 <- NULL
    lib_info <- pData(cds)
    data_df <- data.frame(t(low_dim_coords[c(x, y), ]))
    colnames(data_df) <- c("data_dim_1", "data_dim_2")
    data_df$sample_name <- colnames(cds)
    data_df <- merge(data_df, lib_info, by.x = "sample_name", 
        by.y = "row.names")
    markers_exprs <- NULL
    if (is.null(markers) == FALSE) {
        markers_fData <- subset(fData(cds), gene_short_name %in% 
            markers)
        if (nrow(markers_fData) >= 1) {
            cds_subset <- cds[row.names(markers_fData), ]
            if (cds_subset@expressionFamily@vfamily %in% c("negbinomial", 
                "negbinomial.size")) {
                integer_expression <- TRUE
            }
            else {
                integer_expression <- FALSE
            }
            if (integer_expression) {
                cds_exprs <- exprs(cds_subset)
                if (is.null(sizeFactors(cds_subset))) {
                  stop("Error: to call this function with relative_expr=TRUE, you must call estimateSizeFactors() first")
                }
                cds_exprs <- Matrix::t(Matrix::t(cds_exprs)/sizeFactors(cds_subset))
                cds_exprs <- reshape2::melt(round(as.matrix(cds_exprs)))
            }
            else {
                cds_exprs <- reshape2::melt(as.matrix(exprs(cds_subset)))
            }
            markers_exprs <- cds_exprs
            colnames(markers_exprs)[1:2] <- c("feature_id", "cell_id")
            markers_exprs <- merge(markers_exprs, markers_fData, 
                by.x = "feature_id", by.y = "row.names")
            markers_exprs$feature_label <- as.character(markers_exprs$gene_short_name)
            markers_exprs$feature_label[is.na(markers_exprs$feature_label)] <- markers_exprs$Var1
            markers_exprs$feature_label <- factor(markers_exprs$feature_label, 
                levels = markers[which(markers_exprs$feature_label %in% 
                  markers)])
        }
    }
    if (is.null(markers_exprs) == FALSE && nrow(markers_exprs) > 
        0) {
        data_df <- merge(data_df, markers_exprs, by.x = "sample_name", 
            by.y = "cell_id")
        data_df$value <- with(data_df, ifelse(value >= 0.01, 
            value, NA))
        g <- ggplot(data = data_df, aes(x = data_dim_1, y = data_dim_2)) + 
            facet_wrap(~feature_label)
    }
    else {
        text_df <- data_df %>% dplyr::group_by_(color_by) %>% 
            summarize(text_x = median(x = data_dim_1), text_y = median(x = data_dim_2))
        if (color_by != "Cluster" & !is.numeric(data_df[, color_by])) {
            text_df$label <- paste0(1:nrow(text_df))
            text_df$process_label <- paste0(1:nrow(text_df), 
                "_", as.character(as.matrix(text_df[, 1])))
            process_label <- text_df$process_label
            names(process_label) <- as.character(as.matrix(text_df[, 
                1]))
            #data_df[, color_by] <- process_label[as.character(data_df[, 
            #    color_by])]
        }
        else {
            text_df$label <- as.character(as.matrix(text_df[, 
                1]))
        }
        g <- ggplot(data = data_df, aes(x = data_dim_1, y = data_dim_2))
    }
    if (is.null(markers_exprs) == FALSE && nrow(markers_exprs) > 
        0) {
        if (cds_subset@expressionFamily@vfamily %in% c("negbinomial", 
            "negbinomial.size")) {
            g <- g + plotting_func(aes(color = log10(value + 
                min_expr), alpha = ifelse(!is.na(value), "2", 
                "1")), size = I(cell_size), stroke = I(cell_size/2), 
                na.rm = TRUE) + scale_color_viridis(option = "viridis", 
                name = "log10(values + 0.1)", na.value = "grey80", 
                end = 0.8) + guides(alpha = FALSE) + facet_wrap(~feature_label, 
                nrow = nrow, ncol = ncol)
        }
        else {
            g <- g + plotting_func(aes(color = value, alpha = ifelse(!is.na(value), 
                "2", "1")), size = I(cell_size), stroke = I(cell_size/2), 
                na.rm = TRUE) + scale_color_viridis(option = "viridis", 
                name = "log10(values + 0.1)", na.value = "grey80", 
                end = 0.8) + guides(alpha = FALSE) + facet_wrap(~feature_label, 
                nrow = nrow, ncol = ncol)
        }
    }
    else {
        g <- g + geom_point(aes_string(color = color_by), 
            size = I(cell_size))
        if (show_group_id) {
            g <- g + geom_text(data = text_df, mapping = aes_string(x = "text_x", 
                y = "text_y", label = "label"), size = 4)
        }
    }
    g <- g + monocle:::monocle_theme_opts() + xlab(paste("UMAP", x)) + 
        ylab(paste("UMAP", y)) + theme(legend.position = "top", 
        legend.key.height = grid::unit(0.35, "in")) + theme(legend.key = element_blank()) + 
        theme(panel.background = element_rect(fill = "white")) + 
        theme(text = element_text(size = 15))
    g 
}
```

```{r sup_figure_2}
genes2plot<-fData(dat)[fData(dat)$ASDRiskGene & fData(dat)$FMRPgenes & fData(dat)$isSignificant,]$gene_short_name
geneids2plot<-fData(dat)[fData(dat)$ASDRiskGene & fData(dat)$FMRPgenes & fData(dat)$isSignificant,]$gene_id

p <- monocle::plot_genes_violin(dat[geneids2plot,], grouping="CellType", ncol=3, min_expr=0.1,
		color_by="CellType",panel_order=genes2plot) +
		scale_fill_manual(values=rev(PMcolors)) + theme(legend.position = "top") + monocle:::monocle_theme_opts() + guides(fill=FALSE)

pdf("../plots/Supp_Figure_2_intersection_violin_plots.pdf",width=7,height=8)
p
dev.off()
```

